File: src/main/scala/memctrl/DRAMModel.scala
package memctrl

import chisel3._
import chisel3.util._

class DRAMModelIO extends Bundle {
  // Inputs
  val cs = Input(Bool())
  val ras = Input(Bool())
  val cas = Input(Bool())
  val we = Input(Bool())
  val addr = Input(UInt(32.W))
  val wdata = Input(UInt(32.W))
  
  // Outputs
  val response_complete = Output(Bool())
  val response_data = Output(UInt(32.W))
}

class DRAMModel extends Module {
  val io = IO(new DRAMModelIO())
  
  // DRAM states (not used extensively in this example)
  object DRAMState extends ChiselEnum {
    val DRAM_IDLE, DRAM_ACTIVATE, DRAM_WAIT, DRAM_PRECHARGE = Value
  }
  
  // Timing parameters (delays in cycles)
  val tRCD_DELAY = 5.U  // Row to column delay
  val tCL_DELAY  = 5.U  // CAS latency delay
  val tPRE_DELAY = 10.U // Precharge delay
  val tREFRESH   = 10.U // Refresh delay
  
  // Refresh cycles (for simulation of periodic refresh)
  val REFRESH_CYCLES = 200.U
  
  // Memory storage using a Memory element
  val memory = Mem(1024, UInt(32.W))
  
  // Registers
  val delay_counter        = RegInit(0.U(32.W))
  val refresh_cycle_counter = RegInit(0.U(32.W))
  val memory_activated     = RegInit(0.U(1.W))
  
  // Default outputs
  io.response_complete := false.B
  io.response_data     := 0.U
  
  // Memory corruption flag (for refresh logic)
  val memory_corrupted = RegInit(false.B)
  
  // Tackle Refresh Cycle counter (always counting up)
  refresh_cycle_counter := refresh_cycle_counter + 1.U
  when(refresh_cycle_counter === REFRESH_CYCLES) {
    refresh_cycle_counter := 0.U
    memory_corrupted := true.B  // Mark memory as corrupted (simulate refresh effect)
    memory_activated := 0.U
  }
  
  // For debugging: print current values
  // printf("Core Signal Values: %d %d %d %d\n", io.cs, io.ras, io.cas, io.we)
  // printf("All Signal Values: 0x%x %d\n", io.addr, io.wdata)
  // printf(p"*** [DRAM] Delay Counter: ${delay_counter}. *** \n")
  
  // Compute the next delay value
  val next_delay = WireDefault(delay_counter)
  // printf(p"*** [DRAM] Next Delay Counter: ${next_delay}. *** \n")
  
  // DRAM command processing: use a mutually exclusive branch for each command.
  // In each branch, if the delay counter is zero then load the appropriate delay constant,
  // otherwise decrement the counter. When the counter is about to expire (e.g. equals 1)
  // we complete the command.
  when(io.cs === 1.U) {
    // Command Inhibit: simply reset the delay counter.
    next_delay := 0.U
    io.response_complete := false.B
  } .elsewhen(io.cs === 0.U && io.ras === 0.U && io.cas === 0.U && io.we === 1.U) {
    // Refresh operation
    when(delay_counter === 0.U) {
      next_delay := tREFRESH
      io.response_complete := false.B
      printf("Refreshing: Loaded tREFRESH (%d cycles)\n", tREFRESH)
    } .otherwise {
      next_delay := delay_counter - 1.U
      when(delay_counter === 1.U) {
        // When the delay expires, finish the refresh.
        refresh_cycle_counter := 0.U
        memory_activated := 0.U
        io.response_complete := true.B
      } .otherwise {
        io.response_complete := false.B
      }
    }
  } .elsewhen(io.cs === 0.U && io.ras === 0.U && io.cas === 1.U && io.we === 1.U) {
    // Activate operation
    when(delay_counter === 0.U) {
      next_delay := tRCD_DELAY
      io.response_complete := false.B
      printf("Activating: Loaded tRCD_DELAY (%d cycles)\n", tRCD_DELAY)
    } .otherwise {
      next_delay := delay_counter - 1.U
      when(delay_counter === 1.U) {
        memory_activated := 1.U
        io.response_complete := true.B
        printf("Completed activation.")
      } .otherwise {
        printf("Not completed activation.")
        io.response_complete := false.B
      }
    }
  } .elsewhen(io.cs === 0.U && io.ras === 1.U && io.cas === 0.U && memory_activated === 1.U) {
    // Read/Write operation
    when(delay_counter === 0.U) {
      next_delay := tCL_DELAY
      io.response_complete := false.B
      printf("Read/Write: Loaded tCL_DELAY (%d cycles)\n", tCL_DELAY)
    } .otherwise {
      next_delay := delay_counter - 1.U
      when(delay_counter === 1.U) {
        // On completion, perform the read or write.
        when(io.we === 1.U) {
          // Read operation
          io.response_data := memory.read(io.addr)
          printf("Reading data %d from address 0x%x\n", memory.read(io.addr), io.addr)
        } .otherwise {
          // Write operation
          memory.write(io.addr, io.wdata)
          io.response_data := io.wdata
          printf("[DRAM] Writing %d to address 0x%x\n", io.wdata, io.addr)
        }
        io.response_complete := true.B
      } .otherwise {
        io.response_complete := false.B
      }
    }
  } .elsewhen(io.cs === 0.U && io.ras === 0.U && io.cas === 1.U && io.we === 0.U) {
    // Precharge operation
    when(delay_counter === 0.U) {
      next_delay := tPRE_DELAY
      io.response_complete := false.B
      printf("Precharge: Loaded tPRE_DELAY (%d cycles)\n", tPRE_DELAY)
    } .otherwise {
      next_delay := delay_counter - 1.U
      when(delay_counter === 1.U) {
        memory_activated := 0.U
        io.response_complete := true.B
      } .otherwise {
        io.response_complete := false.B
      }
    }
  } .otherwise {
    next_delay := delay_counter
    io.response_complete := false.B
  }
  
  // Update the delay counter at the clock edge.
  delay_counter := next_delay
}

object DRAMModelMain extends App {
  chisel3.emitVerilog(new DRAMModel(), Array("--target-dir", "generated"))
}

File: behavioral/iteration2/dram_model.cpp
#include "dram_model.h"
#include <iostream>

DRAMModel::DRAMModel()
    : state(DRAM_IDLE), delay_counter(0), current_op(0), current_addr(0), current_data(0), refresh_cycle_counter(0), memory_activated(0) {
}

DRAMModel::~DRAMModel() {
}

void DRAMModel::update(CData cs, CData ras, CData cas, CData we,
                       IData addr, IData wdata,
                       CData* response_complete, IData* response_data) {

    /* Tackle Refresh Logic First */
    refresh_cycle_counter++;

    if(refresh_cycle_counter == REFRESH_CYCLES) { // If we hit the cycle count, we will penalize the memory controller
        refresh_cycle_counter = 0;

        // Corrupt all data
        for(auto &item : memory) {
            memory[item.first] = -1;
        }

        memory_activated = 0;
    }


    // Clear outputs: response_complete is 1-bit; response_data is a 32-bit value.
    *response_complete = 0;
    *response_data = 0;

    // Only consider a command if the DRAM is selected.
    // (Active low: cs, ras, and cas must be 0.)
    if (cs == 1 && ras == 0 && cas == 0) {
        // Determine the operation based on WE.
        // Active low: if we == 0 then it's a write; if we == 1 then it's a read.
        delay_counter = -1;
        // std::cout << "[DRAM] LAUNCHING OPERATION" << std::endl;
    }

    else if(cs == 0 && ras == 0 && cas == 0 && we == 1) {
        if(delay_counter == -1) {
            delay_counter = tREFRESH;
            // std::cout << "[DRAM] LAUNCHING REFRESH OPERATION" << std::endl;
        }

        if (delay_counter > 0) {
            delay_counter--;
            // std::cout << "[DRAM] WAITING FOR REFRESH TO COMPLETE." << std::endl;
            *response_complete = 0;
        }
        else {
            // std::cout << "[DRAM] REFRESHING NOW" << std::endl;
            refresh_cycle_counter = 0;
            memory_activated = 0; 
            delay_counter = -1;
            *response_complete = 1;
        }

    }

    else if(cs == 0 && ras == 0 && cas == 1 && we == 1) {
        if(delay_counter == -1) {
            delay_counter = tRCD_DELAY;
        }

        if (delay_counter > 0) {
            delay_counter--;
            // std::cout << "[DRAM] WAITING FOR TRANSACTION TO COMPLETE." << std::endl;
            *response_complete = 0;
        }
        else {
            // std::cout << "[DRAM] ACTIVATING NOW" << std::endl;
            memory_activated = 1; 
            delay_counter = -1;
            *response_complete = 1;
        }
    }

    // We will not allow the Mem Controller to proceed unless the memory has been activated
    else if (cs == 0 && ras == 1 && cas == 0 && memory_activated == 1) {
        if(delay_counter == -1) {
            delay_counter = tCL_DELAY;
        }
        if (delay_counter > 0) {
            delay_counter--;
            // std::cout << "[DRAM] WAITING FOR TRANSACTION TO COMPLETE." << std::endl;
            *response_complete = 0;
        }
        else if (delay_counter == 0) {
            uint32_t result = 0;
            if (we == 1) {
                // For a read, fetch the stored value (or 0 if not found).
                auto it = memory.find(addr);
                result = (it != memory.end()) ? it->second : 0;
                // std::cout << "[DRAM] Reading from address 0x" << std::hex << addr 
                //           << " - " << std::dec << result << std::endl;
            } else if (we == 0) {
                // For a write, store the value and return it.
                memory[addr] = wdata;
                result = wdata;
                std::cout << "[DRAM] Writing " << result << " to address 0x" << std::hex << addr << std::dec << std::endl;
            } else {
                // std::cout << "[DRAM] Somehow passed through." << std::endl;
            }
            // Produce the 32-bit result.
            *response_data = result;
            *response_complete = 1;
            delay_counter = -1;
        }
    }

    else if(cs == 0 && ras == 0 && cas == 1 && we == 0) {
        if(delay_counter == -1) {
            delay_counter = tPRE_DELAY;
        }
        if (delay_counter > 0) {
            delay_counter--;
            // std::cout << "[DRAM] WAITING FOR PRECHARGE TO COMPLETE." << std::endl;
            *response_complete = 0;
        }
        else if(delay_counter == 0) {
            delay_counter = -1;
            memory_activated = 0;
            *response_complete = 1;
        }
    }
}

void DRAMModel::set_memory(uint32_t addr, uint32_t value) {
    memory[addr] = value;
}
